fastcrypto-zkp/src/bn254/zk_login.rs

```diff
@@ -658,99 +503,154 @@ fn bits_to_bytes(bits: &[bool]) -> Vec<u8> {
     } else {
-        Ok(header)
+        Err(FastCryptoError::GeneralError(format!(
+            "Yet to implement: Unable to hash a vector of length {}",
+            inputs.len()
+        )))
     }
 }
 
-/// Calculate the merklized hash of the given bytes after 0 paddings.
-pub fn calculate_merklized_hash(bytes: &[u8]) -> Result<String, FastCryptoError> {
-    let mut bitarray = bytearray_to_bits(bytes);
-    pad_bitarray(&mut bitarray, 248);
-    let bigints = convert_to_bigints(&bitarray, 248);
-    to_poseidon_hash(bigints)
+/// Convert a bigint string to a field element.
+fn to_field(val: &str) -> Result<Bn254Fr, FastCryptoError> {
+    Bn254Fr::from_str(val).map_err(|_| FastCryptoError::InvalidInput)
 }
 
-/// Calculate the hash of the inputs.
-pub fn to_poseidon_hash(inputs: Vec<Bn254Fr>) -> Result<String, FastCryptoError> {
-    if inputs.len() <= 15 {
-        let mut poseidon1: PoseidonWrapper = PoseidonWrapper::new();
-        Ok(poseidon1.hash(inputs)?.to_string())
-    } else if inputs.len() <= 30 {
-        let mut poseidon1: PoseidonWrapper = PoseidonWrapper::new();
-        let hash1 = poseidon1.hash(inputs[0..15].to_vec())?;
-
-        let mut poseidon2 = PoseidonWrapper::new();
-        let hash2 = poseidon2.hash(inputs[15..].to_vec())?;
+/// Pads a stream of bytes and maps it to a field element
+fn map_bytes_to_field(str: &str, max_size: u16) -> Result<Bn254Fr, FastCryptoError> {
+    if str.len() > max_size as usize {
+        return Err(FastCryptoError::InvalidInput);
+    }
+    let in_arr: Vec<BigUint> = str
+        .chars()
+        .map(|c| BigUint::from_slice(&([c as u32])))
+        .collect();
 
-        let mut poseidon3 = PoseidonWrapper::new();
-        let hash_final = poseidon3.hash([hash1, hash2].to_vec());
+    let str_padded = pad_with_zeroes(in_arr, max_size)?;
+    map_to_field(&str_padded, 8)
+}
 
-        Ok(hash_final?.to_string())
-    } else {
-        Err(FastCryptoError::InvalidInput)
+fn pad_with_zeroes(in_arr: Vec<BigUint>, out_count: u16) -> Result<Vec<BigUint>, FastCryptoError> {
+    if in_arr.len() > out_count as usize {
+        return Err(FastCryptoError::GeneralError("in_arr too long".to_string()));
     }
+    let mut padded = in_arr.clone();
+    padded.extend(vec![
+        BigUint::from_str("0").unwrap();
+        out_count as usize - in_arr.len() as usize
+    ]);
+    Ok(padded)
 }
-/// Convert a bytearray to a bitarray.
-fn bytearray_to_bits(bytearray: &[u8]) -> Vec<bool> {
-    bytearray
-        .iter()
-        .flat_map(|&byte| (0..8).rev().map(move |i| (byte >> i) & 1 == 1))
-        .collect()
+
+/// Maps a stream of bigints to a single field element. First we convert the base from
+/// inWidth to packWidth. Then we compute the poseidon hash of the "packed" input.
+/// input is the input vector containing equal-width big ints. inWidth is the width of
+/// each input element.
+fn map_to_field(input: &[BigUint], in_width: u16) -> Result<Bn254Fr, FastCryptoError> {
```

Comment:
We've decided to rename `map_to_field` to `hash_to_field`..

Reply:
whats "hash" here? the list of big ints?